pkg/mgmt/volume/volume.go

```diff
@@ -96,14 +99,80 @@ func (c *VolController) syncVol(vol *apis.LVMVolume) error {
 		return nil
 	}
 
-	if err = lvm.CreateVolume(vol); err != nil {
-		vol.Status.Error = c.transformLVMError(err)
-		err = lvm.UpdateVolInfo(vol, lvm.LVMStatusFailed)
+	// if there is already a volGroup field set for lvmvolume resource,
+	// we'll first try to create a volume in that volume group.
+	if vol.Spec.VolGroup != "" {
+		err = lvm.CreateVolume(vol)
+		if err == nil {
+			return lvm.UpdateVolInfo(vol, lvm.LVMStatusReady)
+		}
+	}
+
+	vgs, err := c.getVgPriorityList(vol)
+	if err != nil {
 		return err
 	}
 
-	err = lvm.UpdateVolInfo(vol, lvm.LVMStatusReady)
-	return err
+	if len(vgs) == 0 {
+		err = fmt.Errorf("no vg available to serve volume request having regex=%q & capacity=%q",
+			vol.Spec.VgPattern, vol.Spec.Capacity)
+		klog.Errorf("lvm volume %v - %v", vol.Name, err)
+	} else {
+		for _, vg := range vgs {
+			// first update volGroup field in lvm volume resource for ensuring
+			// idempotency and avoiding volume leaks during crash.
+			if vol, err = lvm.UpdateVolGroup(vol, vg.Name); err != nil {
+				klog.Errorf("failed to update volGroup to %v: %v", vg.Name, err)
+				return err
+			}
+			if err = lvm.CreateVolume(vol); err == nil {
+				return lvm.UpdateVolInfo(vol, lvm.LVMStatusReady)
```

Comment:
there could be a leak here if UpdateVolInfo fails, and it pick a new vg for creating the volume. We need to fix it @iyashu

Reply:
I think there will not be any leak here. See we are setting up the vg name first in the spec using `lvm.UpdateVolGroup(vol, vg.Name)` at line 124 above before creating the lv. Lets say `UpdateVolInfo` fails here, in the next sync iteration, you can see we first check if the vg is set in the spec and accordingly call lvm.CreateVolume with same vg and UpdateVolInfo (see the line 104-109). As a result, we'll keep on retrying the `UpdateVolInfo` if it keeps on failing with backoff set by workqueue.

Reply:
sounds good. Thanks @iyashu